### PR TITLE
Telemetry interface implementation

### PR DESCRIPTION
**Pull request by jasoncoolmax**
(for #695)
Implementing the telemetry interface.
A typical usage of the telemetry feature is like:

_[[ADTelemetry getInstance] registerDispatcher:dispatcher aggregationRequired:NO];
[[ADTelemetry getInstance] flush];_

where _dispatcher_ is something developers need to implement, according to protocol _ADDispatcher_, to receive telemetry events.

### CODE REVIEW

File: ADAL/src/ADDefaultDispatcher.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#define GRAB_DISPATCH_LOCK dispatch_semaphore_signal(_dispatchLock)
```
Comment (RPangrle) [inline ADAL/src/ADDefaultDispatcher.h]: If you were to implement a lock using a counting semaphore you'd use wait to grab the lock and signal to release it. However in this case I would recommend just using NSLock

Comment (jasoncoolmax) [inline ADAL/src/ADDefaultDispatcher.h]: Sure :)

File: ADAL/src/public/ADTelemetry.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (RPangrle) [inline ADAL/src/public/ADTelemetry.h]: Please move this header into private for now until we're ready to ship the feature.

File: ADAL/src/public/ADAL.h [hunk 3]
```diff
@@ -58,6 +58,7 @@ typedef void(^ADAuthenticationCallback)(ADAuthenticationResult* result);
 #import <ADAL/ADUserIdentifier.h>
 #import <ADAL/ADUserInformation.h>
 #import <ADAL/ADWebAuthController.h>
+#import <ADAL/ADTelemetry.h>
```
Comment (RPangrle) [inline ADAL/src/public/ADAL.h]: Please remove from umbrella header until we're ready to ship the feature.

File: ADAL/src/ADDefaultDispatcher.m [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        _dispatcher = dispatcher;
+        SAFE_ARC_RETAIN(_dispatcher);
+    }
+    return self;
+}
+
+- (void)flush
+{
+    [_dispatchLock lock]; //avoid access conflict when manipulating _objectsToBeDispatched
+    NSMutableDictionary* objectsToBeDispatchedCopy = _objectsToBeDispatched;
```
Comment (RPangrle) [inline ADAL/src/ADDefaultDispatcher.m]: Note this isn't actually a copy, you might want to rename the local variable.

File: ADAL/src/public/ADAL.h (hunk 3 above)
Comment (jasoncoolmax) [inline ADAL/src/public/ADAL.h]: Sure :)

File: ADAL/src/ADDefaultDispatcher.m (hunk 4 above)
Comment (jasoncoolmax) [inline ADAL/src/ADDefaultDispatcher.m]: How about we simply call it objectToBeDispatched?

File: ADAL/src/ADAggregatedDispatcher.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADDefaultDispatcher.h"
+
+@interface ADAggregatedDispatcher : ADDefaultDispatcher
```
Comment (addev-ashish) [inline ADAL/src/ADAggregatedDispatcher.h]: >ADAggregatedDispatcher [](start = 11, length = 22)

are you using the latest version of the visio?

File: ADAL/src/ADDefaultEvent.m [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    
+    _propertyMap = [[self defaultParameters] mutableCopy];
+    [_propertyMap addObject:@[@"event_name", eventName]];
+    
+    return self;
+}
+
+- (void)setProperty:(NSString*)name value:(NSString*)value
+{
+    if ([NSString adIsStringNilOrBlank:name] || [NSString adIsStringNilOrBlank:value])
```
Comment (addev-ashish) [inline ADAL/src/ADDefaultEvent.m]: >adIsStringNilOrBlank:value [](start = 58, length = 26)

IMO value can be null, do you disagree

File: ADAL/src/ADDefaultEvent.m [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    
+    [_propertyMap addObject:@[name, value]];
+}
+
+- (NSArray*)getProperties
+{
+    return _propertyMap;
+}
+
+- (void)setStartTime:(NSDate*)time
```
Comment (addev-ashish) [inline ADAL/src/ADDefaultEvent.m]: >setStartTime [](start = 8, length = 12)

Any reason why you are moving the start time from telemetry to DefaultEvent?

File: ADAL/src/ADDefaultEvent.m [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    
+    dispatch_once(&s_parametersOnce, ^{
+        
+        s_defaultParameters = [NSMutableArray new];
+        
+#if TARGET_OS_IPHONE
+        //iOS:
+        NSString* deviceId = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
+        NSString* deviceName = [[UIDevice currentDevice] name];
+        SET_IF_NOT_NIL(s_defaultParameters, @"device_id", [[deviceId dataUsingEncoding:NSUTF8StringEncoding] base64EncodedStringWithOptions:0]);
```
Comment (addev-ashish) [inline ADAL/src/ADDefaultEvent.m]: >device_id [](start = 46, length = 9)

i would like to understand how you are setting the device id

File: ADAL/src/ADTelemetry.m [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        else
+        {
+            _dispatcher = [[ADDefaultDispatcher alloc] initWithDispatcher:dispatcher];
+        }
+        SAFE_ARC_RETAIN(_dispatcher);
+    }
+}
+
+- (void)flush
```
Comment (addev-ashish) [inline ADAL/src/ADTelemetry.m]: >flush [](start = 8, length = 5)

not needed anymore.

File: ADAL/src/ADAggregatedDispatcher.h (hunk 5 above)
Comment (jasoncoolmax) [inline ADAL/src/ADAggregatedDispatcher.h]: The "AD" prefix is added to all ADAL class names, so it becomes "ADAggregatedDispatcher" :)

File: ADAL/src/ADDefaultEvent.m (hunk 6 above)
Comment (jasoncoolmax) [inline ADAL/src/ADDefaultEvent.m]: Agree :)

File: ADAL/src/ADDefaultEvent.m (hunk 7 above)
Comment (jasoncoolmax) [inline ADAL/src/ADDefaultEvent.m]: Getting the system current time can be done in either  Event or Telemetry. To make it consistent across platform, I can move it to class Telemetry if we all choose it that way :)

File: ADAL/src/ADDefaultEvent.m (hunk 8 above)
Comment (jasoncoolmax) [inline ADAL/src/ADDefaultEvent.m]: Currently it is using the device's UUID, but seems Mac does not have that. Maybe your solution yesterday should work for both iOS and Mac OS :)

File: ADAL/src/ADTelemetry.m (hunk 9 above)
Comment (jasoncoolmax) [inline ADAL/src/ADTelemetry.m]: I see the word doc has it :P But without flush, when do the cached events be flushed to developer? For both defaultDispatcher and aggregatedDispatcher

**Comment by RPangrle** [comment]
:shipit:

File: ADAL/src/telemetry/ADDefaultDispatcher.m [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        _objectsToBeDispatched = [NSMutableDictionary new];
+        _dispatchLock = [NSLock new];
+        
+        _dispatcher = dispatcher;
+        SAFE_ARC_RETAIN(_dispatcher);
+    }
+    return self;
+}
+
+- (void)flush
```
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADDefaultDispatcher.m]: add a requestId as parameter

File: ADAL/src/telemetry/ADDefaultDispatcher.m [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                [_dispatcher dispatch:properties];
+            }
+        }
+        
+        SAFE_ARC_RELEASE(objectsToBeDispatched);
+    }
+}
+
+- (void)receive:(NSString *)requestId
+          event:(id<ADEventInterface>)event
```
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADDefaultDispatcher.m]: dispatch immediately when we receive.

File: ADAL/src/telemetry/ADAggregatedDispatcher.m [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    return nil;
+}
+
+- (id)initWithDispatcher:(id<ADDispatcher>)dispatcher
+{
+    self = [super initWithDispatcher:dispatcher];
+    return self;
+}
+
+- (void)flush
```
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADAggregatedDispatcher.m]: add requestId as parameter.

**Pull request merged**